lib/gate.h

```diff
@@ -37,43 +58,82 @@ struct Gate {
 
   GateKind kind;
   unsigned time;
-  unsigned num_qubits;
   std::vector<unsigned> qubits;
+  std::vector<unsigned> controlled_by;
+  uint64_t cmask;
   std::vector<fp_type> params;
   Matrix<fp_type> matrix;
-  bool unfusible;      // If true, the gate is fused as a master.
-  bool inverse;        // If true, the qubit order is inversed (q0 > q1).
+  bool unfusible;      // If true, the gate is fused as a parent.
+  bool swapped;        // If true, the qubit order is inversed (q0 > q1).
 };
 
 template <typename Gate, typename GateDef>
 inline Gate CreateGate(unsigned time, unsigned q0,
                        Matrix<typename Gate::fp_type>&& matrix,
                        std::vector<typename Gate::fp_type>&& params = {}) {
-  return Gate{GateDef::kind, time, GateDef::num_qubits, {q0},
-              std::move(params), std::move(matrix), false, false};
+  return Gate{GateDef::kind, time, {q0}, {}, 0,
+              std::forward<std::vector<typename Gate::fp_type>>(params),
+              std::forward<Matrix<typename Gate::fp_type>>(matrix),
+              false, false};
 }
 
 template <typename Gate, typename GateDef>
 inline Gate CreateGate(unsigned time, unsigned q0, unsigned q1,
                        Matrix<typename Gate::fp_type>&& matrix,
                        std::vector<typename Gate::fp_type>&& params = {}) {
-  Gate gate = {GateDef::kind, time, GateDef::num_qubits, {q0, q1},
-               std::move(params), std::move(matrix), false, false};
+  Gate gate = {GateDef::kind, time, {q0, q1}, {}, 0,
+               std::forward<std::vector<typename Gate::fp_type>>(params),
+               std::forward<Matrix<typename Gate::fp_type>>(matrix),
+               false, false};
+
   if (q0 > q1) {
-    gate.inverse = true;
+    gate.swapped = true;
     std::swap(gate.qubits[0], gate.qubits[1]);
-    MatrixShuffle({1, 0}, 2, gate.matrix);
+    if (!GateDef::symmetric) {
+      MatrixShuffle({1, 0}, 2, gate.matrix);
+    }
   }
+
   return gate;
 }
 
 template <typename Gate, typename GateDef>
 inline Gate CreateGate(unsigned time, std::vector<unsigned>&& qubits,
                        Matrix<typename Gate::fp_type>&& matrix = {},
                        std::vector<typename Gate::fp_type>&& params = {}) {
-  return Gate{GateDef::kind, time, static_cast<unsigned>(qubits.size()),
-              std::move(qubits), std::move(params), std::move(matrix),
-              false, false};
+  Gate gate = {GateDef::kind, time,
+               std::forward<std::vector<unsigned>>(qubits), {}, 0,
+               std::forward<std::vector<typename Gate::fp_type>>(params),
+               std::forward<Matrix<typename Gate::fp_type>>(matrix),
+               false, false};
+
+  detail::SortQubits<Gate, GateDef>(gate);
+
+  return gate;
+}
+
+template <typename Gate>
+inline void MakeControlledGate(std::vector<unsigned>&& controlled_by,
```

Comment:
The `MakeControlledGate` methods should probably set `symmetric` to false, since adding control qubits might not preserve gate symmetry.

Reply:
`symmetric` is only used to check if the gate matrix needs to be reshuffled if the gate qubits are not ordered. Adding control qubits does not change the gate matrix and it preserves gate symmetry.